src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,245 @@
+    }
+
+    // If we reach this endpoint, it means the current job is already handling
+    // the loading for the wanted time (same thumbnail).
+    return this._currentJob.jobPromise;
+  }
+
+  private startJob = (contentInfos: IContentInfos,
+                      time: number,
+                      segment: ISegment
+  ): Promise<unknown> => {
+    const killJob$ = new ReplaySubject();
+
+    const abortError$ = killJob$.pipe(
+      map(() => {
+        throw new VideoThumbnailLoaderError("ABORTED",
+                                            "VideoThumbnailLoaderError: Aborted job.");
+      })
+    );
+
+    const jobPromise = observableMerge(
+      initSourceBuffer$(contentInfos,
+                        this._videoElement).pipe(
+        mergeMap((videoSourceBuffer) => {
+          const removeBuffers$: Observable<unknown> =
+            removeBufferForTime$(this._videoElement, videoSourceBuffer, time);
+          return removeBuffers$.pipe(
+            mergeMap(() => {
+              log.debug("VTL: Removed buffer before appending segments.", time);
+
+              return segmentLoader({
+                manifest: contentInfos.manifest,
+                period: contentInfos.manifest.periods[0],
+                adaptation: contentInfos.adaptation,
+                representation: contentInfos.representation,
+                segment,
+              }).pipe(
+                filter((evt): evt is { type: "data";
+                                       value: { responseData: Uint8Array }; } =>
+                  evt.type === "data"),
+                mergeMap((evt) => {
+                  const inventoryInfos = { manifest: contentInfos.manifest,
+                                           period: contentInfos.period,
+                                           adaptation: contentInfos.adaptation,
+                                           representation: contentInfos.representation,
+                                           segment };
+                  return videoSourceBuffer
+                    .pushChunk({ data: { chunk: evt.value.responseData,
+                                         timestampOffset: 0,
+                                         appendWindow: [undefined, undefined],
+                                         initSegment: null,
+                                         codec: contentInfos
+                                           .representation.getMimeTypeString() },
+                                 inventoryInfos })
+                      .pipe(map(() => {
+                        log.debug("VTL: Appended segment.", evt.value.responseData);
+                        this._videoElement.currentTime = time;
+                        if (this._videoElement.buffered.length === 0) {
+                          throw new VideoThumbnailLoaderError("NOT_BUFFERED",
+                                                              "No buffered data after loading.");
```

Comment:
Sometimes browsers are weird and buffered seems to not be immediately updated. Less not throw in that case.

Also it's a little random to just base ourselves on buffered.length (what if there already was data?)

Reply:
Ok